src/coord/src/coord.rs

```diff
@@ -3277,6 +3277,20 @@ impl Coordinator {
             session.add_transaction_ops(TransactionOps::Tail)?;
         }
 
+        let source_id = match expr {
+            TailExpr::Id(id) => id,
+            TailExpr::View { name, view } => {
+                let (ops, index_id) =
+                    self.generate_view_ops(session, name.clone(), view, None, false)?;
+                assert!(index_id.is_none());
+                self.catalog_transact(ops, |_builder| Ok(())).await?;
+                self.catalog
+                    .try_get(&name, session.conn_id())
+                    .expect("must exist")
+                    .id()
+            }
```

Comment:
I think we can frob the `DataflowBuilder` to avoid this! Try this on for size?

1. Hoist the dataflow builder creation up to L3279 (`let mut builder = self.dataflow_builder();`).
2. In the `TailExpr::View` case, call `allocate_transient_id` to get a fresh global ID, then call `builder.import_view_into_dataflow` using that transient ID with the optimized view as the view argument.
3. Return that transient ID as the `source_id` for use below.

If that works you can ditch the `name` in `TailExpr::View` too!

Reply:
Noodling with this and it needs a mapping from id to a CatalogEntry, which has to be in the catalog state. To get there we need to run catalog_transact and pass a catalog::Op::CreateItem that still wants a FullName and other stuff, so we'd still end up with a leftover entry there? That is all needed by determine_frontier and nearest_indexes.

I'm also really unclear on what to do with the dataflow that is passed to import_view_into_dataflow. I made a new dataflow right before the import call and now am confused. Calling ship_dataflow on it seems wrong because then it would need to be dropped, but also the ship_dataflow below for the sink will take care of building whatever it needs. And in any case the sink dataflow build still needs a mapping from the id to the mir, which still means we need an entry in the CatalogState?

Am I missing something?

Reply:
Do we make a copy of the CatalogState, directly call insert_item on it and make a DataflowBuilder from that when building the sink? Still unsure why we need the call to import_view_into_dataflow at all because I don't know what to do with its dataflow.

Reply:
Ah sorry, you reached the limits of my ability to describe this without actually trying it. I pushed up another commit that seems to do the job. I fixed the TopK thing too while I was in the area.

Reply:
Your explanation sounded like it was a thing that could be done in a few lines, and I thought I was missing something. Happy to see it took more work I guess? Ha, and thanks.

Reply:
Heh, yeah, `DataflowBuilder` supported all the right pieces but figuring out how to share code in `handle_sequence_tail` was not trivial.